File: hybrid/ex1.f90
! This is a simple hybrid hello world program.
! Prints MPI information
! For each task/thread
!  task id
!  node name for task
!  thread id
!  # of threads for the task
!  core on which the thread is running

module getit
contains
! Get the core on which a thread is running
  function get_core_c()
      USE ISO_C_BINDING, ONLY: c_long, c_char, C_NULL_CHAR, c_int
      use numz
      implicit none
      integer(in8) :: get_core_c
      interface
         integer(c_long) function cfunc() BIND(C, NAME='sched_getcpu')
            USE ISO_C_BINDING, ONLY: c_long, c_char
         end function cfunc
      end interface
      get_core_c = cfunc()
   end function
end module

program hybrid
    use getit
    implicit none
    include 'mpif.h'
    integer numtasks,myid,ierr
    character (len=MPI_MAX_PROCESSOR_NAME):: myname
    character(len=MPI_MAX_LIBRARY_VERSION_STRING+1) :: version
    integer mylen,vlan,mycore
    integer OMP_GET_MAX_THREADS,OMP_GET_THREAD_NUM
    call MPI_INIT( ierr )
    call MPI_COMM_RANK( MPI_COMM_WORLD, myid, ierr )
    call MPI_COMM_SIZE( MPI_COMM_WORLD, numtasks, ierr )
    call MPI_Get_processor_name(myname,mylen,ierr)
! print the MPI libraty version
if (myid .eq. 0)then
      write(*,*)" Fortran MPI TASKS ",numtasks
      call MPI_Get_library_version(version, vlan, ierr)
      write(*,*)trim(version)
    endif
!$OMP PARALLEL
!$OMP CRITICAL
    mycore=get_core_c()
    write(unit=*,fmt="(a,i4,a,a)",advance="no") &
                " task ",myid, " is running on ",trim(myname)
    write(unit=*,fmt="(a,i4,a,i4,a,i4)") &
            " thread= ",OMP_GET_THREAD_NUM(), &
            " of ",OMP_GET_MAX_THREADS(),     &
            " is on core ",mycore
!$OMP END CRITICAL
!$OMP END PARALLEL
    call MPI_FINALIZE(ierr)
end program

File: hybrid/ex1.c
#include <stdio.h>
#include <mpi.h>
#include <omp.h>
int sched_getcpu();
/************************************************************
This is a simple hybrid hello world program.
Prints MPI information 
For each task/thread
  task id
  node name for task
  thread id
  # of threads for the task
  core on which the thread is running
************************************************************/
int main(int argc, char **argv)
{
    int myid,numtasks,resultlen;
    char version[MPI_MAX_LIBRARY_VERSION_STRING];
    char myname[MPI_MAX_PROCESSOR_NAME] ;
    int vlan;
    int myt,nt,mycore;
    MPI_Init(&argc,&argv);
    MPI_Comm_size(MPI_COMM_WORLD,&numtasks);
    MPI_Comm_rank(MPI_COMM_WORLD,&myid);
    MPI_Get_processor_name(myname,&resultlen); 
    if (myid == 0 ) {
	    printf(" C MPI TASKS %d\n",numtasks);
	    MPI_Get_library_version(version, &vlan);
	    printf("%s\n",version);
    }
#pragma omp parallel
  {
    nt = omp_get_num_threads();
    myt= omp_get_thread_num();
    mycore=sched_getcpu();
#pragma omp critical
    printf("task %d running on %s  thread %d of %d on core %d\n",
		    myid,myname,myt,nt,mycore);
  }
    MPI_Finalize();
}

